### PR TITLE
Refactor long tuples to records in typeclass.ml

### PR DESCRIPTION
**Pull request by ulugbekna**
Addressing issue #7927. 

I refactored only four functions that take the same long tuple. (Short commits for easier reviewing) 
Those four functions are `extract_type_decls`, `merge_type_decls`, `final_env`, and `check_coercions`. 
The long tuple was replaced by a record type that I created `'a class_res`. The type's name and its fields' names should probably be changed. 

Should that type also be revealed in the interface file? Afaik, `'a class_res` is only used by those four functions internal to the `typeclass.ml`.

### CODE REVIEW

**Comment by ulugbekna** [comment]
All `typing-objects` tests passed. However, `typing-objects-bugs` tests were not in the log for some unknown-to-me reason. The tests failed during `lib-bigarray-2` testing, but that shouldn't be in relation to my work.

**Comment by ulugbekna** [comment]
Next I would want to work on tuple `(cl, id, ty_id, obj_id, obj_params, obj_ty, cl_id, cl_params, cl_ty, constr_type, dummy_class)` used in several places in `typeclass.ml`, e.g. `class_infos` function.

File: typing/typeclass.ml
Comment (dra27) [inline typing/typeclass.ml]: Have a care committing TODO comments - it'll be cleaner in the final history if they're not there at all (I personally find `git add -p` more helpful than `git add -u`, but it depends on how pernickety you are with your commit-crafting fu!)

File: typing/typeclass.ml
Comment (dra27) [inline typing/typeclass.ml]: As per https://github.com/ocaml/ocaml/issues/7927#issuecomment-474041486, it would be good in this refactoring to be making use of punning wherever possible - so arguments like this could be renamed to `clsres_id` and then the "transforming" (e.g. from `cl.pci_name` to `clsres_id_loc`) becomes more visually obvious.

File: typing/typeclass.ml
Comment (dra27) [inline typing/typeclass.ml:1677]: Definitely use punning here

**Review by dra27 — COMMENTED** [review]
This looks like a nice start, thanks! Two related thoughts:

 - Try to maximize the use of field punning - i.e. be allergic to writing `{ foo = bar }` if a rename can give `{ foo }` instead
 - I would think for these new types (especially ones internal to a module) that we can drop the `clsres_` prefix on the field names and simply require a type annotation at the site. i.e. I think that `let foo ({id; id_loc; clty} : 'a class_res) = …` is nicer than `let foo {clsres_id; clsres_id_loc; clsres_clty} = …` and definitely nicer than `let foo {clsres_id = id; clsres_id_loc = id_loc; clsres_clty = clty} = …`. I can't remember if this is something which used not to be possible or whether this kind of a prefixing is just our lpdwHungarian notation hangover from pre-Merlin days!

**Comment by gasche** [comment]
I think it's okay to have a small scope for a first PR on this topic; settling on record field names for this record and code style for those functions. We can have further PRs that deal with other functions later, even if that means changing the present record to break it in smaller pieces.

I agree with @dra27's suggestions (drop the prefix and use punning as much as possible).

File: typing/typeclass.ml
Comment (ulugbekna) [reply typing/typeclass.ml]: Thank you for taking time to review my PR! Would you recommend from committing TODOs even if they are removed before the merge?

File: typing/typeclass.ml
Comment (ulugbekna) [inline typing/typeclass.ml]: I was thinking if we could come up with a more expressive name for this record type.

File: typing/typeclass.ml
Comment (ulugbekna) [inline typing/typeclass.ml]: Would `decls` be a better naming than `cls_res`?

File: typing/typeclass.ml
Comment (dra27) [reply typing/typeclass.ml]: I'd recommend against putting them in the same commits as other things, unless your intention is to squash everything together at the end.

File: typing/typeclass.ml
Comment (dra27) [reply typing/typeclass.ml]: The name of the function I think is already a bit strange (not yours, I realise) - it has a plural name, but it's actual a singular operation used with `map2` further down. Perhaps rename it `decl`?

File: typing/typeclass.ml
Comment (dra27) [reply typing/typeclass.ml]: The name isn't actually used, so I wouldn't worry about it too much! I'm not familiar with the type checker - isn't this record ultimately the result of typing an `object` expression (as opposed to `Typing.class_type_declaration` which is for the type of the `class` itself). In which case might `'a object_expression` or `'a object_declaration` work?

Comment (gasche) [reply typing/typeclass.ml]: I'm not sure either (maybe @lpw25 would know), but from looking at the way this data is used in the module I would still call this `class_decl` or `class_declaration`. The producer of this record is `final decl`, which I understand as returning the "final version of the declaration", called by the `type_classes` function which deals with a mutually-recursive block of class declarations (according to its name and the name of its callers).

Comment (lpw25) [reply typing/typeclass.ml]: > isn't this record ultimately the result of typing an object expression 

Nope, this is for the full class or class type. (By the way "object expression" is normally called either "class signature"/"class structure" to distinguish it from "object literal" -- which is when a class structure is used directly as an expression.)

**Comment by gasche** [comment]
(I'm happy to let @dra27 continue to drive the review for this PR, approving when satisfied with the state.)

File: typing/typeclass.ml
Comment (dra27) [inline typing/typeclass.ml]:
```suggestion
    let self_type = Ctype.self_type clty.cty_type in
```
I'd recommend installing the Git `pre-commit` hook detailed in [CONTRIBUTING.md](https://github.com/ocaml/ocaml/blob/trunk/CONTRIBUTING.md#coding-guidelines) (I'm biased, as I wrote it!)

File: typing/typeclass.ml
Comment (dra27) [inline typing/typeclass.ml]: A style suggestion, now that punning is in place, would be to put all the punned identifiers together and reduce the line count):

```ocaml
   { id; clty; ty_id; cltydef; obj_id; obj_abbr; cl_id; cl_abbr; arity;
     pub_meths; coe; expr;
     id_loc = cl.pci_name;
```

(to me it draws more attention to the fields which require expressions - the type checker has your back for any missing fields, of course)

File: typing/typeclass.ml
Comment (dra27) [inline typing/typeclass.ml]: Similar comment as before on the line-count

File: typing/typeclass.ml
Comment (gasche) [reply typing/typeclass.ml]: (I use `./tools/check-typo-since trunk`. Is this all properly documented in HACKING.adoc?)

**Comment by ulugbekna** [comment]
Sorry it took me so long; had a midterms week. Let me know what you think about the fix. 

Btw, I copied the pre-commit hook to my .git and configured my editor to automatically remove trailing whitespace. Thanks, @dra27 :-)

File: typing/typeclass.ml
Comment (ulugbekna) [reply typing/typeclass.ml]: @gasche coding guidelines and `tools/check-typo` are nicely documented in [CONTRIBUTING.md](https://github.com/ocaml/ocaml/blob/trunk/CONTRIBUTING.md#coding-guidelines). My bad not checking it out first.

**Comment by dra27** [comment]
This LGTM - it's an internal refactoring, so up to you if you'd like to push a `Changes` entry.

**Comment by ulugbekna** [comment]
That's a little internal refactoring. I'd say it would rather clutter `Changes` than add valuable info, right?

**Comment by gasche** [comment]
We have an `Internal/compiler-libs change` section in the changelog for internal changes. Having entries for new contributors is nice, so please add a Changes entry there. (If you were to do further refactoring PRs on this file, you could just add them to the same entry instead of having more and more entries.)

**Comment by ulugbekna** [comment]
We don't need to put a label, e.g. `GPR`, before the issue number anymore, right? So, I have 
`- #7927: Replace long tuples into records in typeclass.ml
  (Ulugbek Abdullaev, review by David Allsopp and Gabriel Scherer)`?

**Comment by Octachron** [comment]
Indeed, however you should also add the pr number:
```
- #7927, #8527: Replace long tuples into records in typeclass.ml
  (Ulugbek Abdullaev, review by David Allsopp and Gabriel Scherer)
```

**Review by gasche — APPROVED** [review]
It is an omission on our part to not have merged this PR earlier. I will rebase to fix the conflict, push and merge when the CI passes.

**Pull request merged**
